### PR TITLE
MGMT-4739: AgentServiceConfig controller cleanup

### PR DESCRIPTION
**Pull request by djzager**
The original PR#1422 left some comments unaddressed. This commit
addresses those comments and other concerns:

- Moving the passwordGenerator to common.go
- Making the error message when NAMESPACE isn't specified more clear
- Documenting that building the bundle+index isn't required
- Simplifying controller constants
- Abstracting the newFilesystemPVC+newDatabasePVC -> newPVC
- Emitting an event when reconcile completes
- Handle OPENSHIFT_VERSIONS
- Set IPV6_SUPPORT to True on assisted-service
- Set requests on assisted-service+postgres containers
- Add some basic guidelines for the filesystem+database PVC requests
- Add SERVICE_ACCOUNT_NAME using d/w API to operator, pass that to
  the assisted-service-deployment
- Add 'http://' to host for `SERVICE_BASE_URL`

### CODE REVIEW

File: internal/controller/api/v1alpha1/agentserviceconfig_types.go [hunk 1]
```diff
@@ -26,10 +26,15 @@ import (
 type AgentServiceConfigSpec struct {
 	// FileSystemStorage defines the spec of the PersistentVolumeClaim to be
 	// created for the assisted-service's filesystem (logs, etc).
+	// With respect to the resource requests, the amount of filesystem storage
+	// consumer will depend largely on the number of clusters you intend to create.
+	// It's best to request as much as possible with the minimum being 20GiB in
+	// development/testing and 100Gib anywhere else.
```
Comment (flaper87) [inline internal/controller/api/v1alpha1/agentserviceconfig_types.go:32]: I got these numbers from my testing environment, which may help with providing a more accurate indication: 

* ~110 MB per cluster (this includes ignition, install-config, and other files as well as the minimal-iso)
* ~2.2 GB per supported version (see all the rhcos files `rhcos-47.83.202102090044-0-minimal.iso  rhcos-47.83.202102090044-0.initrd.img   rhcos-47.83.202102090044-0.iso          rhcos-47.83.202102090044-0.rootfs.img   rhcos-47.83.202102090044-0.vmlinuz` )


Here's the screenshot

DeepinScreenshot_select-area_20210409130338

File: internal/controller/api/v1alpha1/agentserviceconfig_types.go [hunk 2]
```diff
@@ -26,10 +26,15 @@ import (
...
 	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Storage for service filesystem"
 	FileSystemStorage corev1.PersistentVolumeClaimSpec `json:"filesystemStorage"`
 	// DatabaseStorage defines the spec of the PersistentVolumeClaim to be
 	// created for the database's filesystem.
+	// With respect to the resource requests, 10Gi is sufficient.
```
Comment (flaper87) [inline internal/controller/api/v1alpha1/agentserviceconfig_types.go:37]: It would be nice to use the current size of the DB in `cloud.redhat.com` as a reference. I've asked how big it is already, waiting on an reply

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 3]
```diff
@@ -430,25 +399,25 @@ func (r *AgentServiceConfigReconciler) newAssistedServiceDeployment(instance *ad
 		{Name: "SKIP_CERT_VERIFICATION", Value: "True"},
 
 		// database
-		newSecretEnvVar("DB_HOST", "db.host", databaseSecretName),
-		newSecretEnvVar("DB_NAME", "db.name", databaseSecretName),
-		newSecretEnvVar("DB_PASS", "db.password", databaseSecretName),
-		newSecretEnvVar("DB_PORT", "db.port", databaseSecretName),
-		newSecretEnvVar("DB_USER", "db.user", databaseSecretName),
+		newSecretEnvVar("DB_HOST", "db.host", databaseName),
+		newSecretEnvVar("DB_NAME", "db.name", databaseName),
```
Comment (flaper87) [inline internal/controller/controllers/agentserviceconfig_controller.go:403]: FWIW, I'm not a huge fan of using the default `postgres` database. I'd rather have a dedicated DB for the service.

File: cmd/operator/main.go [hunk 4]
```diff
@@ -34,7 +34,8 @@ import (
 )
 
 const (
-	NamespaceEnvVar string = "NAMESPACE"
+	NamespaceEnvVar         string = "NAMESPACE"
+	OpenshiftVersionsEnvVar string = "OPENSHIFT_VERSIONS"
```
Comment (flaper87) [inline cmd/operator/main.go:38]: perhaps we could move this const to `internal/controller/controllers/images.go` so we can use it in both places.

File: internal/controller/controllers/agentserviceconfig_controller.go [hunk 5]
```diff
@@ -493,7 +461,12 @@ func (r *AgentServiceConfigReconciler) newAssistedServiceDeployment(instance *ad
 				MountPath: "/data",
 			},
 		},
-		Resources: corev1.ResourceRequirements{},
+		Resources: corev1.ResourceRequirements{
+			Requests: corev1.ResourceList{
+				corev1.ResourceCPU:    resource.MustParse("300m"),
+				corev1.ResourceMemory: resource.MustParse("400Mi"),
+			},
```
Comment (flaper87) [inline internal/controller/controllers/agentserviceconfig_controller.go:468]:
```suggestion
			Requests: corev1.ResourceList{
				corev1.ResourceCPU:    resource.MustParse("200m"),
				corev1.ResourceMemory: resource.MustParse("512Mi"),
			},
```

These request values seem to be a bit low. Here's what I see from `kubectl top pod` in my test env:

```
[dev@edge-10 dev-scripts]$ kubectl top pod
NAME                               CPU(cores)   MEMORY(bytes)
assisted-service-94f447959-8fdgq   15m          517Mi
ocp-metal-ui-d5df44bd5-2s8v8       0m           5Mi
postgres-7ccd794b8-dbdrv           30m          104Mi
```

AI was idle when I ran this command and there were 2 cluster records. 

P.S: One could argue that using `512Mi` when the service is idle sounds like there may be a leak or some aggressive, `in-memory` caching.

**Review by flaper87 — CHANGES_REQUESTED** [review]
small changes requested

some tests would be nice, tho :stuck_out_tongue:

File: internal/controller/api/v1alpha1/agentserviceconfig_types.go (hunk 2 above)
Comment (djzager) [reply internal/controller/api/v1alpha1/agentserviceconfig_types.go:37]: I probably also shouldn't be saying that 10Gi is sufficient....maybe that it's the minimum recommended.

File: internal/controller/controllers/agentserviceconfig_controller.go (hunk 3 above)
Comment (djzager) [reply internal/controller/controllers/agentserviceconfig_controller.go:403]: not dealing with this here.

File: internal/controller/controllers/agentserviceconfig_controller.go (hunk 5 above)
Comment (djzager) [reply internal/controller/controllers/agentserviceconfig_controller.go:468]: updated the requests...but I'm not looking into the possibility of a memory leak

File: cmd/operator/main.go [hunk 6]
```diff
@@ -80,7 +82,25 @@ func main() {
...
+	// this prevents us from having to include the full json in source
+	// ie. this should ALWAYS be specified on the CSV, until a proper
+	// API is provided for it
+	// I think it's reasonable to check that the OPENSHIFT_VERSIONS is
+	// legit before we go passing it down to the assisted-service deployment
+	// and letting it fail there.
+	var openshiftVersionsMap models.OpenshiftVersions
+	openshiftVersions, found := os.LookupEnv(controllers.OpenshiftVersionsEnvVar)
+	if !found || openshiftVersions == "" {
+		setupLog.Error(fmt.Errorf("%s environment variable must be set (commonly set automatically in every Pod) to a non-empty value.", controllers.OpenshiftVersionsEnvVar), "unable to get OpenShift Versions")
```
Comment (rollandf) [inline cmd/operator/main.go:99]: NIIT the error message seems not related "commonly set automatically in every Pod"

**Review by rollandf — APPROVED** [review]

File: cmd/operator/main.go (hunk 6 above)
Comment (djzager) [reply cmd/operator/main.go:99]: That's specifically referring to what we've done with the manager.yaml so that it's automatically included in the ClusterServiceVersion. I'll spend some time thinking about if there are ways to make that more clear.
